File: lib/onchain/scripts/js/get_signers.mjs
#!/usr/bin/env node
/**
 * Get signers from GCDecryptionVerifier contract.
 * Reads the contract address from GCDecryptionVerifierAddress.sol and calls getSigners().
 * Writes the signers to a file.
 */
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';
import Web3 from 'web3';
import { SodaWeb3Helper } from './soda-web3-helper.mjs';

function extractAddressFromSolFile(filePath) {
    try {
        const content = fs.readFileSync(filePath, 'utf8');
        
        // Match: address constant GCDecryptionVerifierAddress = 0x...;
        const pattern = /address\s+constant\s+\w+\s*=\s*(0x[a-fA-F0-9]+);/;
        const match = content.match(pattern);
        
        if (match) {
            const address = match[1];
            return Web3.utils.toChecksumAddress(address);
        } else {
            throw new Error(`Could not find address constant in ${filePath}`);
        }
    } catch (error) {
        if (error.code === 'ENOENT') {
            throw new Error(`Address file not found: ${filePath}`);
        }
        throw error;
    }
}

async function getSignersFromContract(sodaHelper, contractAddress) {
    // Setup the contract (compile to get ABI)
    const contractPath = "core/contracts/";
    
    const success = await sodaHelper.setupContract(
        contractPath,
        "GCDecryptionVerifier.sol",
        "decryption_verifier",
        false, // overwrite
        "", // MpcInterfacePath
        "", // mpcCorePath
        "" // address_path
    );
    
    if (!success) {
        throw new Error("Failed to set up the contract");
    }
    
    // Set the contract address
    const contract = sodaHelper.getContract("decryption_verifier");
    if (!contract) {
        throw new Error("Contract not found after setup");
    }
    contract.options.address = contractAddress;
    
    // Call getSigners() view function
    const signers = await sodaHelper.callContractView("decryption_verifier", "getSigners", []);
    
    if (!signers) {
        throw new Error("Failed to call getSigners()");
    }
    
    return signers;
}

function writeSignersToFile(signers, outputFile) {
    // Create directory if it doesn't exist
    const dir = path.dirname(outputFile);
    if (dir && dir !== '.') {
        fs.mkdirSync(dir, { recursive: true });
    }
    
    let content = "# Signers from GCDecryptionVerifier contract\n";
    content += `# Total signers: ${signers.length}\n\n`;
    
    signers.forEach((signer, index) => {
        content += `${index + 1}. ${signer}\n`;
    });
    
    fs.writeFileSync(outputFile, content, 'utf8');
    console.log(`Signers written to ${outputFile}`);
}

export async function getSignersAddresses() {
    // Get environment variables
    const signingKey = process.env.SIGNING_KEY;
    if (!signingKey) {
        throw new Error("SIGNING_KEY environment variable not set");
    }
    
    const providerUrl = process.env.REMOTE_RPC_URL;
    if (!providerUrl) {
        throw new Error("REMOTE_RPC_URL environment variable not set");
    }
    
    const chainId = process.env.REMOTE_CHAIN_ID;
    if (!chainId) {
        throw new Error("REMOTE_CHAIN_ID environment variable not set");
    }
    
    let chainName = process.env.CHAIN_NAME;
    if (!chainName) {
        throw new Error("CHAIN_NAME environment variable not set");
    }
    // Convert to uppercase to match Python script behavior
    chainName = chainName.toUpperCase();
    
    // Get address file path
    const addressFileChain = `lib/onchain/contracts/${chainName}/GCDecryptionVerifierAddress.sol`;
    let contractAddress = null;
    
    if (fs.existsSync(addressFileChain)) {
        contractAddress = extractAddressFromSolFile(addressFileChain);
    } else {
        throw new Error(`Address file not found. Checked:\n  - ${addressFileChain}`);
    }
    
    // Initialize SodaWeb3Helper
    const sodaHelper = new SodaWeb3Helper(signingKey, providerUrl, parseInt(chainId));
    
    // Get signers
    const signers = await getSignersFromContract(sodaHelper, contractAddress);
    
    return signers;
}

// Main execution - run if this file is executed directly
// Check if this is the main module by comparing the file path
const mainModulePath = process.argv[1];
const currentModulePath = fileURLToPath(import.meta.url);

if (mainModulePath && (mainModulePath === currentModulePath || mainModulePath.endsWith('get_signers.mjs'))) {
    (async () => {
        try {
            const signers = await getSignersAddresses();
            console.log(`\nSuccessfully retrieved ${signers.length} signer(s)`);
            process.exit(0);
        } catch (error) {
            console.error(`Error: ${error.message}`);
            console.error(error.stack);
            process.exit(1);
        }
    })();
}

File: lib/onchain/scripts/python/get_signers.py
#!/usr/bin/env python3
"""
Get signers from GCDecryptionVerifier contract.
Reads the contract address from GCDecryptionVerifierAddress.sol and calls getSigners().
Writes the signers to a file.
"""
import os
import sys
import re
from web3 import Web3
from lib.onchain.scripts.python.soda_web3_helper import SodaWeb3Helper

def extract_address_from_sol_file(file_path):
    """Extract the address constant from a Solidity file."""
    try:
        with open(file_path, 'r') as f:
            content = f.read()
        
        # Match: address constant GCDecryptionVerifierAddress = 0x...;
        pattern = r'address\s+constant\s+\w+\s*=\s*(0x[a-fA-F0-9]+);'
        match = re.search(pattern, content)
        
        if match:
            address = match.group(1)
            return Web3.to_checksum_address(address)
        else:
            raise ValueError(f"Could not find address constant in {file_path}")
    except FileNotFoundError:
        raise FileNotFoundError(f"Address file not found: {file_path}")

def get_signers_from_contract(soda_helper, contract_address):
    """Get signers from the GCDecryptionVerifier contract."""
    # Setup the contract (compile to get ABI)
    contract_path = "contracts/GCDecryptionVerifier.sol"
    success = soda_helper.setup_contract(
        contract_path, 
        "decryption_verifier",
        mpc_inst_path="",
        mpc_core_path="",
        address_path=""
    )
    
    if not success:
        raise Exception("Failed to set up the contract")
    
    # Set the contract address
    soda_helper.set_contract_address("decryption_verifier", contract_address)
    
    # Call getSigners() view function
    signers = soda_helper.call_contract_view("decryption_verifier", "getSigners", [])
    
    return signers

def write_signers_to_file(signers, output_file):
    """Write signers to a file."""
    os.makedirs(os.path.dirname(output_file) if os.path.dirname(output_file) else '.', exist_ok=True)
    
    with open(output_file, 'w') as f:
        f.write("# Signers from GCDecryptionVerifier contract\n")
        f.write(f"# Total signers: {len(signers)}\n\n")
        for i, signer in enumerate(signers, 1):
            f.write(f"{i}. {signer}\n")
    
    print(f"Signers written to {output_file}")

def get_signers_addresses():
    """Main function."""
    # Get environment variables
    signing_key = os.environ.get('SIGNING_KEY')
    if not signing_key:
        raise ValueError("SIGNING_KEY environment variable not set")
    
    provider_url = os.environ.get('REMOTE_RPC_URL')
    if not provider_url:
        raise ValueError("REMOTE_RPC_URL environment variable not set")
    
    chain_id = os.environ.get('REMOTE_CHAIN_ID')
    if not chain_id:
        raise ValueError("REMOTE_CHAIN_ID environment variable not set")

    chain_name = os.environ.get('CHAIN_NAME')
    if not chain_name:
        raise ValueError("CHAIN_NAME environment variable not set")
    chain_name = chain_name.upper()
    
    # Get address file path 
    address_file_chain = f"contracts/GCDecryptionVerifierAddress.sol"
    contract_address = None
    if os.path.exists(address_file_chain):
        contract_address = extract_address_from_sol_file(address_file_chain)
    else:
        raise FileNotFoundError(f"Address file not found. Checked:\n  - {address_file_chain}")
    
    # Initialize SodaWeb3Helper
    soda_helper = SodaWeb3Helper(signing_key, provider_url, chain_id=int(chain_id))
    
    # Get signers
    signers = get_signers_from_contract(soda_helper, contract_address)
    
    return signers

if __name__ == "__main__":
    try:
        signers = get_signers_addresses()
        print(f"\nSuccessfully retrieved {len(signers)} signer(s)")
        sys.exit(0)
    except Exception as e:
        print(f"Error: {e}", file=sys.stderr)
        import traceback
        traceback.print_exc()
        sys.exit(1)

